### PR TITLE
Update SSO to use 4.4 discovery and refresh tokens

### PR DESCRIPTION
**Pull request by OskarDamkjaer**

### CODE REVIEW

File: src/shared/modules/connections/connectionsDuck.ts [hunk 1]
```diff
@@ -513,6 +519,7 @@ export const startupConnectEpic = (action$: any, store: any) => {
       // merge with discovery data if we have any and try again
       if (discovered) {
         store.dispatch(discovery.updateDiscoveryConnection(discovered))
+        console.log('discovered: ' + JSON.stringify(discovered))
```
Comment (jharris4) [inline src/shared/modules/connections/connectionsDuck.ts:522]: console.log to be removed here?

File: src/shared/modules/connections/connectionsDuck.ts [hunk 2]
```diff
@@ -532,7 +539,8 @@ export const startupConnectEpic = (action$: any, store: any) => {
                 store.dispatch(setActiveConnection(discovery.CONNECTION_ID))
                 resolve({ type: STARTUP_CONNECTION_SUCCESS })
               })
-              .catch(() => {
+              .catch(err => {
+                console.log('error: ' + JSON.stringify(err))
```
Comment (jharris4) [inline src/shared/modules/connections/connectionsDuck.ts:543]: console.log to be removed here?

File: src/shared/modules/connections/connectionsDuck.ts [hunk 3]
```diff
@@ -630,7 +638,66 @@ export const connectionLostEpic = (action$: any, store: any) =>
     .filter(() => inWebEnv(store.getState()) && isConnected(store.getState()))
     .throttleTime(5000)
     .do(() => store.dispatch(updateConnectionState(PENDING_STATE)))
-    .mergeMap(() => {
+    .mergeMap((action: any) => {
+      if (action.error.code === 'Neo.ClientError.Security.TokenExpired') {
```
Comment (jharris4) [inline src/shared/modules/connections/connectionsDuck.ts:642]: we're planning to refactor so this code is not duplicated, right?

File: src/shared/modules/connections/connectionsDuck.ts [hunk 4]
```diff
@@ -630,7 +638,66 @@ export const connectionLostEpic = (action$: any, store: any) =>
...
+        const SSOProviders = getActiveConnectionData(store.getState())
+          ?.SSOProviders
+        if (SSOProviders) {
+          return Rx.Observable.fromPromise(
+            (async function() {
+              //TODO can throw?
```
Comment (jharris4) [inline src/shared/modules/connections/connectionsDuck.ts:648]: yeah, if the `handleRefreshingToken` throws, we should bail out of the reconnection attempts, and go straight to the connection lost state. TBD is whether we should specifically report this back to the user somehow, possible with custom error text?

File: src/shared/modules/discovery/discoveryDuck.ts [hunk 5]
```diff
@@ -164,111 +158,57 @@ export const discoveryOnStartupEpic = (some$: any, store: any) => {
...
-
-      if (action.forceURL) {
-        const { username, protocol, host } = getUrlInfo(action.forceURL)
-
-        const discovered = {
-          username,
-          requestedUseDb: action.requestedUseDb,
-          host: `${protocol ? `${protocol}//` : ''}${host}`,
-          supportsMultiDb: !!action.requestedUseDb,
-          encrypted: action.encrypted,
```
Comment (jharris4) [inline src/shared/modules/discovery/discoveryDuck.ts:188]: I'm a bit fuzzy on while all these discovered properties are needed, and whether they're actually used anywhere, but I guess it's fine for now.

File: src/shared/modules/connections/connectionsDuck.ts (hunk 1 above)
Comment (OskarDamkjaer) [reply src/shared/modules/connections/connectionsDuck.ts:522]: done

File: src/shared/modules/connections/connectionsDuck.ts (hunk 3 above)
Comment (OskarDamkjaer) [reply src/shared/modules/connections/connectionsDuck.ts:642]: yes

File: src/shared/modules/discovery/discoveryHelpers.test.ts [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    expect(discoveryData).toBeTruthy()
+    expect(discoveryData?.host).toEqual(boltHost)
+    expect(discoveryData?.source).toEqual(DISCOVERY_ENDPOINT)
+    expect(discoveryData?.SSOProviders).toEqual([])
+    expect(discoveryData?.SSOError).toEqual(undefined)
+    expect(discoveryData?.neo4jVersion).toEqual(neo4jVersion)
+    expect(discoveryData?.urlMissing).toEqual(false)
+  })
+
+  test('finds and priotises sso providers from session storage/connect form when all discovery sources are present, but doesnt merge when hosts differ', async () => {
```
Comment (jharris4) [inline src/shared/modules/discovery/discoveryHelpers.test.ts:90]: should probably add the new discoveryConnection source to this test?

File: src/shared/modules/discovery/discoveryDuck.ts (hunk 5 above)
Comment (OskarDamkjaer) [reply src/shared/modules/discovery/discoveryDuck.ts:188]: Yeah this turned into a bit of a rabbit hole trying to understand where they're all used. I'm somewhat certain that the encrypted key is not used here and that the restApi is used for the :http command. But didn't have enough time to dig into it fully

**Review by jharris4 — APPROVED** [review]
This is working really well in testing so far, so I'm going to merge it so we can have a release candidate.

We may want to address some of the minor remaining review comments later though...

**Pull request merged**